mobile/src/main/java/ch/epfl/sdp/mobile/ui/ar/ArScreen.kt

```diff
@@ -32,42 +60,107 @@ fun ArScreen(modifier: Modifier = Modifier) {
+
+    val boardBoundingBox = board!!.modelInstance?.filamentAsset?.boundingBox!!
+
+    // get height (on y axe) of the board
+    // Double the value to get the total height of the box
+    val boardYOffset = 2 * boardBoundingBox.halfExtent[1]
+    val boardHalfSize = boardBoundingBox.halfExtent[0]
+
+    arBoard = ArBoard(BoardBorderSize, boardYOffset, boardHalfSize)
+
+    // DEBUG
+    if (DisplayAxes) {
+      white = loadModel("models/white.glb", nodePlacementMode)
+      red = loadModel("models/red.glb", nodePlacementMode)
+      blue = loadModel("models/blue.glb", nodePlacementMode)
+      green = loadModel("models/green.glb", nodePlacementMode)
+    }
   }
 
   AndroidView(
       factory = { ArSceneView(it) },
-      modifier = modifier,
+      modifier = modifier.testTag("AR"),
       update = { view ->
-        addNode(pawn1, view)
-        addNode(pawn2, view)
+        if (DisplayAxes) {
+          white?.let {
+            it.scale(2f)
+            board?.addChild(it)
+            it.placementPosition = Position(y = arBoard!!.boardHeight)
+          }
+          red?.let {
+            it.scale(2f)
+            board?.addChild(it)
+            it.placementPosition = Position(x = 3f, y = arBoard!!.boardHeight)
+          }
+          blue?.let {
+            it.scale(2f)
+            board?.addChild(it)
+            it.placementPosition = Position(z = 3f, y = arBoard!!.boardHeight)
+          }
+          green?.let {
+            it.scale(2f)
+            board?.addChild(it)
+            it.placementPosition = Position(y = 3f + arBoard!!.boardHeight)
+          }
+        }
+
+        /** Add the given [piece] on the board in the correct position */
+        fun addPiece(piece: ArModelNode) {
+          piece.let {
+            board?.addChild(it)
+            it.placementPosition =
+                arBoard?.toArPosition(piecePosition) ?: ArModelNode.DEFAULT_PLACEMENT_POSITION
+          }
+        }
+
+        /**
+         * TODO : With the [ChessBoardState], iterate over the list of pieces, and do the same that
+         * [pawn1]
+         */
+        if (pawn1 != null) {
+          addPiece(pawn1!!)
+        }
+
+        // Scale down the board size
+        // As all the pieces are the board children, they scale as well
+        board?.scale(BoardScale)
+
+        /**
+         * If not already in the scene, the board will be added. Update the board anchor with the
+         * given one and change the displayed position
+         *
+         * @param anchor The (new) board's anchor position
+         */
+        fun anchorOrMoveBoard(anchor: Anchor) {
+          // Add only one instance of the node
+          if (!view.children.contains(board!!)) {
+            view.addChild(board!!)
+          }
+          board?.anchor = anchor
+        }
+
+        // Place the board on the taped position
+        view.onTouchAr = { hitResult, _ -> anchorOrMoveBoard(hitResult.createAnchor()) }
```

Comment:
Question : what happens if the callback is called before `board` is loaded ? Could the `board` be null and the app crash ?

Reply:
Indeed, if the board is null, the app crash at line 157. We have a null safety on the `board`.